### ISSUE TITLE
css3 layout

### ISSUE THREAD

**Issue by RachelComerford** [w3c/publ-cg#31]
(from BISG survey)

- What are best practices for applying CSS3 in layout of epub3

**Comment by dauwhe** [comment]
CSS3 is not a meaningful term. Some CSS modules are labeled as level 3, as they extend the features of CSS 2. But many new features are level 1. There are level 4 and level 5 drafts. The level doesn't really mean much. 

If the question is about newer CSS features that were not officially supported in EPUB 3.0.1, then it is a very, very general question. We might need to divide this up into specific features. Grid? Flex? New display types? New font features? Variables? Calc?

**Comment by JayPanoz** [comment]
Yeah it will indeed be super difficult to keep up with the whole, especially as modules are now implemented and usable in production within 2 years – it could be 5–7 years before e.g. flex.

Then, when it comes to layout modules, I would probably set them apart from fonts, custom props, the `calc()` function, etc. 

My reasoning is it’s almost guaranteed new layout modules will have fragmentation-specific issues – so browser vendors might not even know those issues exist if we don’t report them. 

Why using fragmentation instead of columns, you may ask? Cos’ it’s not just about columns: print media and the `setPagination` API in iOS – a superset of columns which behaves very differently – share a lot of those issues. So columns are only the tip of the iceberg and the real fixes are to be implemented at the fragmentation’s logic, which is a tricky subject (cf. [this discussion about CSS Regions turning into a fragmentation discussion at some point](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/kTktlHPJn4Q/YrnfLxeMO7IJ)).

I agree with @dauwhe there, dividing may be the best option + non-layout modules will be low-hanging fruits compared to layout modules.

**Comment by llemeurfr** [comment]
Such a discussion must take into account a reality that seems too often hidden: they are two kinds of reading systems: those which are built around a Web rendering engine (Chrome, Webkit) and therefore support the extended range of CSS properties supported by the underlying engine; and on the other side almost every e-ink reader and some well known apps (incl. FBReader I think) which are built around an in-house rendering engine and support a limited range of CSS properties (from different levels). 
It would be more than sad to have best practices limited to the intersection of all RS implementations, but it's very very difficult to create EPUB 3 publications that contain "progressive enhancements".

**Comment by teytag** [comment]
Hi Le Meur,

I agree with you. When designing the EPUB3 format, we recommend only one application for the customer Android and IOS. When more RS requests are made, there is a problem. As customers can not get the performance they are cooling down from EPUB3 and returning to PDF.

Best regards

N. Erhan Uzumcu
Art Director and EPUB Designer

**Comment by llemeurfr** [comment]
... and this is, we know it, the main reason why EPUB 2 is still alive: it is basic therefore it works potentially on any RS. We won't have EPUB 2 out until we face this situation. 

We could create a "basic" EPUB 3 profile (with limited HTML5 and CSS functionalities) described by best practices, usable on nearly any RS and tagged using a metadata property; but the risk would be that authors limit themselves to this profile. 

After all, is this our main risk, compared to the risk of seing EPUB 2 alive for the next 20 years? but I digress.

**Comment by RachelComerford** [comment]
This is a helpful conversation - thanks! I know very little about CSS and was really only able to relay the comment from the survey without providing much additional information. I do have a contact email address from the contributor of this query from Gutenberg and if someone on this chain is interested in taking ownership of the initial documentation for this, I can put you in touch to hash out where specifically they are looking for guidance.

**Comment by dauwhe** [comment]
This might also be something that will ultimately be split into many different best practices :)

**Comment by JayPanoz** [comment]
I just left the group and will consequently be unable to keep working on this issue. But please contact me to pick up where I left off – sorry, DGPR in Europe so I can’t post email adresses publicly and in theory, I shouldn’t even pass them privately.
